torch/csrc/Exceptions.h

```diff
@@ -360,33 +367,52 @@ struct PyWarningHandler {
     in_exception_ = true;
   }
 
+  // Trigger processing of warnings
+  TORCH_PYTHON_API void process_warnings();
+
  private:
   InternalHandler internal_handler_;
   at::WarningHandler* prev_handler_;
   bool in_exception_;
 };
 
 namespace detail {
+
+struct noop_gil_scoped_release {
+  noop_gil_scoped_release() {
+    // suppress `unused variable` error messages at call sites
+    (void)(this != (this + 1));
+  }
+};
+
+template <bool release_gil>
+using conditional_gil_scoped_release = std::conditional_t<
+    release_gil,
+    pybind11::gil_scoped_release,
+    noop_gil_scoped_release>;
+
 template <typename Func, size_t i>
 using Arg = typename invoke_traits<Func>::template arg<i>::type;
 
-template <typename Func, size_t... Is>
+template <bool release_gil, typename Func, size_t... Is>
 auto wrap_pybind_function_impl_(
     Func&& f,
     std::index_sequence<Is...>,
-    bool release_gil) {
+    std::bool_constant<release_gil>) {
   using result_type = typename invoke_traits<Func>::result_type;
   namespace py = pybind11;
 
   // f=f is needed to handle function references on older compilers
-  return [f = std::forward<Func>(f),
-          release_gil](Arg<Func, Is>... args) -> result_type {
+  return [f = std::forward<Func>(f)](Arg<Func, Is>... args) -> result_type {
     HANDLE_TH_ERRORS
-    if (release_gil) {
-      py::gil_scoped_release no_gil;
-      return c10::guts::invoke(f, std::forward<Arg<Func, Is>>(args)...);
+    conditional_gil_scoped_release<release_gil> no_gil;
+    if constexpr (std::is_void_v<result_type>) {
```

Comment:
Why is this needed now, can't the else pattern always be used?

Reply:
Because `void foo = ...` invalid even if the `void` is deduced. There is a proposal to make `void` a regular type which would allow this but it isn't there yet.

Reply:
Interesting limitation. Sounds good!